src/vulkan/pipeline.cc

```diff
@@ -65,87 +65,107 @@ void Pipeline::Shutdown() {
   if (r.IsSuccess())
     command_->SubmitAndReset(fence_timeout_ms_);
 
-  DestoryDescriptorPools();
-  DestoryDescriptorSetLayouts();
+  for (size_t desc = 0; desc < descriptor_sets_.size(); ++desc) {
+    vkDestroyDescriptorSetLayout(device_, descriptor_sets_[desc].layout,
+                                 nullptr);
+    if (descriptor_sets_[desc].empty)
+      continue;
+
+    vkDestroyDescriptorPool(device_, descriptor_sets_[desc].pool, nullptr);
+  }
+
   command_->Shutdown();
   vkDestroyPipelineLayout(device_, pipeline_layout_, nullptr);
   for (size_t i = 0; i < descriptors_.size(); ++i)
     descriptors_[i]->Shutdown();
   vkDestroyPipeline(device_, pipeline_, nullptr);
 }
 
-void Pipeline::DestoryDescriptorSetLayouts() {
-  for (size_t i = 0; i < descriptor_set_layouts_.size(); ++i) {
-    vkDestroyDescriptorSetLayout(device_, descriptor_set_layouts_[i], nullptr);
-  }
-  descriptor_set_layouts_.clear();
+void Pipeline::SortDescriptorsBySetAndBinding() {
+  if (!need_sort_descriptors_)
+    return;
+
+  std::sort(descriptors_.begin(), descriptors_.end(),
+            [](const std::unique_ptr<Descriptor>& a,
+               const std::unique_ptr<Descriptor>& b) {
+              if (a->GetDescriptorSet() == b->GetDescriptorSet())
+                return a->GetBinding() < b->GetBinding();
+              return a->GetDescriptorSet() < b->GetDescriptorSet();
+            });
+  need_sort_descriptors_ = false;
 }
 
 Result Pipeline::CreateDescriptorSetLayouts() {
-  std::sort(descriptors_.begin(), descriptors_.end());
+  SortDescriptorsBySetAndBinding();
 
   size_t i = 0;
+  uint32_t next_desc = 0;
   while (i < descriptors_.size()) {
     std::vector<VkDescriptorSetLayoutBinding> bindings;
-    for (const uint32_t current_desc = descriptors_[i]->GetDescriptorSet();
-         i < descriptors_.size() &&
-         current_desc == descriptors_[i]->GetDescriptorSet();
-         ++i) {
+    const uint32_t current_desc = descriptors_[i]->GetDescriptorSet();
+    while (i < descriptors_.size() &&
```

Comment:
What is the relation between descriptors_ and descriptor_sets_? Is it 1:1?

Reply:
N:1
Each element of `descriptor_sets_` can be matched to one or more elements of `descriptors_`